src/core/int128.d

```diff
@@ -0,0 +1,696 @@
+/* 128 bit integer arithmetic.
+ *
+ * Not optimized for speed.
+ *
+ * Copyright: Copyright D Language Foundation 2022.
+ * License:   $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
+ * Authors:   Walter Bright
+ * Source:    $(DRUNTIMESRC core/_int128.d)
+ */
+
+module core.int128;
+
+nothrow:
+@safe:
+@nogc:
+
+alias I = long;
+alias U = ulong;
+enum Ubits = U.sizeof * 8;
+
+struct Cent
+{
+    U lsl;      // low 64 bits
+    U msl;      // high 64 bits
+}
+
+enum One = Cent(1);
+enum Zero = Cent();
+
+/*****************************
+ * Test against 0
+ * Params:
+ *      c = Cent to test
+ * Returns:
+ *      true if != 0
+ */
+pure
+bool tst(Cent c)
+{
+    return c.msl || c.lsl;
+}
+
+
+/*****************************
+ * Complement
+ * Params:
+ *      c = Cent to complement
+ * Returns:
+ *      complemented value
+ */
+pure
+Cent com(Cent c)
+{
+    c.lsl = ~c.lsl;
+    c.msl = ~c.msl;
+    return c;
+}
+
+/*****************************
+ * Negate
+ * Params:
+ *      c = Cent to negate
+ * Returns:
+ *      negated value
+ */
+pure
+Cent neg(Cent c)
+{
+    return inc(com(c)); // ~c + 1
+}
+
+/*****************************
+ * Increment
+ * Params:
+ *      c = Cent to increment
+ * Returns:
+ *      incremented value
+ */
+pure
+Cent inc(Cent c)
+{
+    return add(c, One);
+}
+
+/*****************************
+ * Decrement
+ * Params:
+ *      c = Cent to decrement
+ * Returns:
+ *      incremented value
+ */
+pure
+Cent dec(Cent c)
+{
+    return sub(c, One);
+}
+
+/*****************************
+ * Shift left one bit
+ * Params:
+ *      c = Cent to shift
+ * Returns:
+ *      shifted value
+ */
+pure
+Cent shl1(Cent c)
+{
+    c.msl = (c.msl << 1) | (cast(I)c.lsl < 0);
+    c.lsl <<= 1;
+    return c;
+}
+
+/*****************************
+ * Unsigned shift right one bit
+ * Params:
+ *      c = Cent to shift
+ * Returns:
+ *      shifted value
+ */
+pure
+Cent shr1(Cent c)
+{
+    c.lsl = (c.lsl >> 1) | ((c.msl & 1) << (Ubits - 1));
+    c.msl >>= 1;
+    return c;
+}
+
+
+/*****************************
+ * Arithmetic shift right one bit
+ * Params:
+ *      c = Cent to shift
+ * Returns:
+ *      shifted value
+ */
+pure
+Cent sar1(Cent c)
+{
+    c.lsl = (c.lsl >> 1) | ((c.msl & 1) << (Ubits - 1));
+    c.msl = cast(I)c.msl >> 1;
+    return c;
+}
+
+/*****************************
+ * Shift left n bits
+ * Params:
+ *      c = Cent to shift
+ *      n = number of bits to shift
+ * Returns:
+ *      shifted value
+ */
+pure
+Cent shl(Cent c, uint n)
+{
+    if (n >= Ubits * 2)
+        return Zero;
+
+    foreach (i; 0 .. n)
+        c = shl1(c);
```

Comment:
```
c.lsl = ((c.lsl << n) | (c.msl >> (Ubits - n - 1) >> 1));
c.msl = c.msl << n;
```
(Untested)

Reply:
I used a lot of reliance on other functions to get many working. No attempt was made to optimize them. Many of my initiative PRs are currently stalled, which disincentivizes me to spend a lot of effort tuning them. So I went for the baseline implementation.

> (Untested)

I know, it doesn't work :-)

P.S. the unittests I wrote uncovered several embarrassing bugs in it.

Reply:
> > (Untested)
> 
> I know, it doesn't work :-)
> 
> P.S. the unittests I wrote uncovered several embarrassing bugs in it.

Ah, yes, because I got the field names the wrong way round, which begs the question, why now just name them `low` and `high` to prevent all confusion.